### PR TITLE
chore(optimizer): check equivalent plans

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Check equivalent plans after applying an optimizer rule.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/heuristic_optimizer.rs [hunk 1]
```diff
@@ -81,6 +88,32 @@ impl<'a> HeuristicOptimizer<'a> {
...
+        for (a, b) in input_plan
+            .schema()
+            .fields
+            .iter()
+            .zip_eq_fast(output_plan.schema().fields.iter())
+        {
+            if a.data_type != b.data_type {
+                fail();
+            }
+        }
```
Comment (liurenjie1024) [inline src/frontend/src/optimizer/heuristic_optimizer.rs:115]: Can we just derive `Eq` for `Schema`?

**Review by liurenjie1024 — APPROVED** [review]
Generally LGTM

File: src/frontend/src/optimizer/heuristic_optimizer.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/heuristic_optimizer.rs:115]:
```rust
#[derive(Clone, PartialEq, Eq, Hash)]
pub struct Field {
    pub data_type: DataType,
    pub name: String,
    /// For STRUCT type.
    pub sub_fields: Vec<Field>,
    /// The user-defined type's name, when the type is created from a protobuf schema file,
    /// this field will store the message name.
    pub type_name: String,
}
```

Comment (liurenjie1024) [reply src/frontend/src/optimizer/heuristic_optimizer.rs:115]: I don't rhonk so, name is part of schema

**Pull request merged**